google/earthenterprise
# Move load and save code into storage manager (with segfault fixes)

This reintroduces the code from PR #1390 to pull the load and save code into the storage manager with fixes to prevent segfaults on CentOS 6 and 7. Many of the changes in this PR are the same as the changes in the previous PR and thus have already been reviewed. It may be helpful to look at the commits individually:

- [Commit cb2864f](https://github.com/google/earthenterprise/commit/cb2864fafbbd47495506b18a0ed0b982de88782b) reintroduces the changes from PR #1390, so all of this code has been reviewed previously.
- [Commit 2a21847](https://github.com/google/earthenterprise/commit/2a21847f21facb2135cb36f1b66300e9315345be) fixes the segmentation fault. This is the most important change to review.
- [Commit 7f5f0bb](https://github.com/google/earthenterprise/commit/7f5f0bbaf2b7466bedf663bf8515a48d784d7065) is some cleanup that mostly involves removing unused functions.

It is not immediately clear why this fix works. My best guess is that the previous version of the code triggered a bug in old versions of GCC and this fix avoids it.

To test:
1. Download the tutorial data and extract it to `/gevol/src`
1. Create the directories `Tutorial/Resources/Vector` and `Tutorial/MapLayers` under the asset root.
1. Run the following script and observe that system manager does not segfault:
```
#!/bin/bash

/opt/google/bin/genewvectorresource --sourcedate 2003-05-10 --encoding UTF-8 --layer 0 --provider USGS-M -o Tutorial/Resources/Vector/CAHighways /gevol/src/Vector/california_roads_line.shp
/opt/google/bin/genewmaplayer --legend=CA_Roads  --default_state_on --output=Tutorial/MapLayers/CaRoads --layericon=road --template=/gevol/src/template/CA_Freeways_MapLayer.kmdsp Tutorial/Resources/Vector/CAHighways
/opt/google/bin/gebuild Tutorial/MapLayers/CaRoads
```
